### PR TITLE
kvserver: don't apply ExcludeDataFromBackup outside config bounds

### LINKED ISSUE #120122: exclude_data_from_backup may erroneously be applied to newly-created table
**Describe the problem**

`exclude_data_from_backup` is a property that can be set on a table to exclude it from backups. This is useful for some system tables and for user tables that are used for things like transient queues.

`exclude_data_from_backup` is part of the span configuration. When evaluating an export request, the _replica_ decides whether it's data should be excluded from backup.  This poses a problem for newly created tables as there is a window in which they may still be part of a replica whose span configuration is from a table with `exclude_data_from_backup` set.

**Reproduction**

*Setup*

- Create a table
- Set exclude_data_from_backup on it
- Wait for it's range to split. It should now be the "last" range.

NB: We turn off implicit transactions because in the next step we want to run statements very quickly after each other but not in the same transaction.

```
./cockroach demo --multitenant=false --empty
demo@127.0.0.1:26257/defaultdb> set enable_implicit_transaction_for_batch_statements=off;
demo@127.0.0.1:26257/defaultdb> create table foo(pk int primary key);
demo@127.0.0.1:26257/defaultdb> alter table foo set (exclude_data_from_backup = true)
demo@127.0.0.1:26257/defaultdb> show cluster ranges;
... SNIP ...
  /Table/63               | /Table/64               |       65 | {1}      | {"region=us-east1,az=b"} | {1}             | {}                  | {}               | NULL
  /Table/64               | /Table/65               |       66 | {1}      | {"region=us-east1,az=b"} | {1}             | {}                  | {}               | NULL
  /Table/65               | /Table/104              |       67 | {1}      | {"region=us-east1,az=b"} | {1}             | {}                  | {}               | NULL
  /Table/104              | /Max                    |       68 | {1}      | {"region=us-east1,az=b"} | {1}             | {}                  | {}               | NULL
(68 rows)
```

Now, if we create a table and then _very quickly_ try to fingerprint it. We often get a 0 valued fingerprint. Implying that this table that has a value has no value. This is because fingerprint is based on ExportRequest

```
demo@127.0.0.1:26257/defaultdb> create table foo2(pk int primary key); insert into foo2 values (1); select crdb_internal.fingerprint(crdb_internal.table_span('foo2'::regclass::oid::int), true);                                                                                                                                                                       
CREATE TABLE
INSERT 0 1
  crdb_internal.fingerprint
-----------------------------
                          0
(1 row)
```

If we wait a few moments for the range to split, we get a non-zero fingerprint:

```
demo@127.0.0.1:26257/defaultdb> select crdb_internal.fingerprint(crdb_internal.table_span('foo2'::regclass::oid::int), true);                                                                                                                                                                                                                                           
  crdb_internal.fingerprint
-----------------------------
       -2919453955530127916
(1 row)
```

We can further prove this to ourselves by now trying this again now that the "last" range is one that doesn't have exclude_data_from_backup on it. We immediately get a non-zero fingerprint result:

```
demo@127.0.0.1:26257/defaultdb> create table foo3(pk int primary key); insert into foo3 values (1); select crdb_internal.fingerprint(crdb_internal.table_span('foo3'::regclass::oid::int), true);                                                                                                                                                                       
CREATE TABLE
INSERT 0 1
  crdb_internal.fingerprint
-----------------------------
       -2919453955530127916
(1 row)
```

Jira issue: CRDB-36495

### PR DESCRIPTION
**Pull request by stevendanna**
Some span configuration updates result in a replica split. However, between the new span configuration being set on the replica and the split happening, the span configuration stored on the replica may be incorrect for a portion of the replica's keyspace.

This is a general problem with asyncronous span configuration application. But, here we introduce a specific fix for the ExcludeDataFromBackup span configuration item.

When setting the span configuration, we now also store the bounds of that span configuration. ExcludeDataFromBackup only returns true if the data being considered for exclusion is contained by those bounds.

Further, we only set DataExcludedFromBackup to true in
BatchTimestampBeforeGCError if the span of the batch that generated
the error is contained by the span configuration bound. This field is
used to ignore BatchTimestampBeforeGCErrors. As a result, not setting
it to true could result in BACKUP failures until the range
splits. However, the alternative is silently ignoring a possible real
error.

Note that this doesn't solve all problems related to DataExcludedFromBackup. For instance, span configuration propagation is asyncronous and thus data for a given table may continue to be excluded from backups for some period after `ALTER TABLE <TABLE> SET (exclude_data_from_backup = false)` has been set.

Fixes #120122

Release note (bug fix): Fix a rare condition in which a BACKUP taken shortly after ALTER TABLE <TABLE> SET (exclude_data_from_backup = true) could result in data from an unrelated table being excluded from a backup.

### CODE REVIEW

File: pkg/kv/kvserver/asim/state/impl.go [hunk 1]
```diff
@@ -1245,6 +1245,21 @@ func (s *state) GetSpanConfigForKey(
 	return *rng.config, nil
 }
 
+// GetSpanConfigForKeyWithBounds is added for the spanconfig.StoreReader interface, required for
+// SpanConfigConformanceReport.
```
Comment (stevendanna) [inline pkg/kv/kvserver/asim/state/impl.go:1249]:
```suggestion
// GetSpanConfigForKeyWithBounds is added for the spanconfig.StoreReader interface.
```

File: pkg/config/system.go [hunk 2]
```diff
@@ -339,9 +339,19 @@ func (s *SystemConfig) getZoneConfigForKey(
 func (s *SystemConfig) GetSpanConfigForKey(
 	ctx context.Context, key roachpb.RKey,
 ) (roachpb.SpanConfig, error) {
+	config, _, err := s.GetSpanConfigForKeyWithBounds(ctx, key)
+	return config, err
+}
+
+// GetSpanConfigForKeyWithBounds looks of the span config for the given
+// key. It's part of spanconfig.StoreReader interface. Note that it is only
+// usable for the system tenant config.
```
Comment (stevendanna) [inline pkg/config/system.go:348]:
```suggestion
// GetSpanConfigForKeyWithBounds returns the span config for the given
// key and the bounds that span the configuration applies to. It's part of spanconfig.StoreReader interface. Note that it is only
// usable for the system tenant config.
```

**Review by arulajmani — APPROVED** [review]
:lgtm: modulo nits

> Reviewed 25 of 25 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120188#-:-NstFASX2hiW3gloMnm4:b-1p9zor)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dt and @stevendanna)

___
*[`pkg/config/system.go` line 348 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120188#gh-1523766195:-NstCMXC7ifdY7vFTNWV:b-9d63k7) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ea6bcb3e33ca9743fdefec488f245c987ad1c2/pkg/config/system.go#L348)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

```suggestion
// GetSpanConfigForKeyWithBounds returns the span config for the given
// key and the bounds that span the configuration applies to. It's part of spanconfig.StoreReader interface. Note that it is only
// usable for the system tenant config.
```
</blockquote></details>

Updated comment looks good to me
___
*[`pkg/kv/kvserver/replica.go` line 1230 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120188#-NstDTGCATF_7LNes7SL:-NstDTGCATF_7LNes7SM:busbmk4) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ea6bcb3e33ca9743fdefec488f245c987ad1c2/pkg/kv/kvserver/replica.go#L1230)):*
> ```Go
> func (r *Replica) excludeReplicaFromBackupRLocked(rspan roachpb.RSpan) bool {
> 	if r.mu.conf.ExcludeDataFromBackup {
> 		// If ExcludeDataFromBackup is set, we also want to ensure that
> ```

nit: consider de-duplicating some of the logic shared here and above.
___
*[`pkg/spanconfig/spanconfig.go` line 292 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120188#-NstEgDOA-HgVwaNiEff:-NstEgDOA-HgVwaNiEfg:bl846th) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ea6bcb3e33ca9743fdefec488f245c987ad1c2/pkg/spanconfig/spanconfig.go#L292)):*
> ```Go
> 	ComputeSplitKey(ctx context.Context, start, end roachpb.RKey) (roachpb.RKey, error)
> 	GetSpanConfigForKey(ctx context.Context, key roachpb.RKey) (roachpb.SpanConfig, error)
> 	GetSpanConfigForKeyWithBounds(ctx context.Context, key roachpb.RKey) (roachpb.SpanConfig, roachpb.Span, error)
> ```

nit: instead of having these two methods, should we instead have `GetSpanConfigForKey` return the span it applies to as well? Callers that don't care about the span are free to ignore it.

Separately, consider adding a comment here and/or explicitly naming the second return value to something like "appliesTo".

**Review by stevendanna — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120188)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani and @dt)

___
*[`pkg/kv/kvserver/replica.go` line 1230 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120188#-NstDTGCATF_7LNes7SL:-NtLKzlS0dgXZEMxgpLj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ea6bcb3e33ca9743fdefec488f245c987ad1c2/pkg/kv/kvserver/replica.go#L1230)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: consider de-duplicating some of the logic shared here and above.
</blockquote></details>

Done.
___
*[`pkg/spanconfig/spanconfig.go` line 292 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120188#-NstEgDOA-HgVwaNiEff:-NtLKu1nBhcrVRNmhjS4:bx4nh62) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ea6bcb3e33ca9743fdefec488f245c987ad1c2/pkg/spanconfig/spanconfig.go#L292)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: instead of having these two methods, should we instead have `GetSpanConfigForKey` return the span it applies to as well? Callers that don't care about the span are free to ignore it.

Separately, consider adding a comment here and/or explicitly naming the second return value to something like "appliesTo".
</blockquote></details>

Done and done.

**Comment by stevendanna** [comment]
bors r=arulajmani

**Pull request merged**